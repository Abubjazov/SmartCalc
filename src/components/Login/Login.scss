@import '../../variables';

.login-widget {
	margin-top: 30vh;
	display: grid;
	grid-template-columns: auto;
	grid-auto-rows: 25px;
	row-gap: 15px;
	justify-items: center;

	a {
		margin-top: 30px;
		border-radius: 20px;
		padding-top: 11px;
		width: 100px;
		height: 40px;
		text-align: center;
		font-size: 18px;
		line-height: 18px;
		background-color: $main-color;
		color: $white;
		text-decoration: none;
		transition: all 0.1s linear;
	}

	input {
		height: 25px;
		width: 260px;
		border: none;
		border-bottom: 2px solid $main-color;
	}
}
