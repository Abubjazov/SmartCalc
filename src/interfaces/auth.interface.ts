export enum AuthActionTypes {
	FETCH_TOKEN = 'FETCH_TOKEN',
	FETCH_TOKEN_SUCCESS = 'FETCH_TOKEN_SUCCESS',
	FETCH_TOKEN_ERROR = 'FETCH_TOKEN_ERROR',

	DELETE_TOKEN = 'DELETE_TOKEN',
	DELETE_TOKEN_SUCCESS = 'DELETE_TOKEN_SUCCESS',
	DELETE_TOKEN_ERROR = 'DELETE_TOKEN_ERROR',

	CLEAR_ERROR = 'CLEAR_ERROR',
}

export interface AuthData {
	email: string
	password: string
}

export interface UserData {
	token: string
	name: string
	email: string
}

export interface AuthState {
	token: null | string
	name: string
	email: string
	status: 'waiting' | 'error' | 'loading'
	error: null | string
}

interface FetchTokenAction {
	type: AuthActionTypes.FETCH_TOKEN
	payload: AuthData
}

interface FetchTokenSuccessAction {
	type: AuthActionTypes.FETCH_TOKEN_SUCCESS
	payload: UserData
}

interface FetchTokenErrorAction {
	type: AuthActionTypes.FETCH_TOKEN_ERROR
	payload: string
}

interface DeleteTokenAction {
	type: AuthActionTypes.DELETE_TOKEN
}

interface DeleteTokenSuccessAction {
	type: AuthActionTypes.DELETE_TOKEN_SUCCESS
}

interface DeleteTokenErrorAction {
	type: AuthActionTypes.DELETE_TOKEN_ERROR
	payload: string
}

interface ClearErrorAction {
	type: AuthActionTypes.CLEAR_ERROR
}

export type AuthAction =
	| FetchTokenAction
	| FetchTokenSuccessAction
	| FetchTokenErrorAction
	| DeleteTokenAction
	| DeleteTokenSuccessAction
	| DeleteTokenErrorAction
	| ClearErrorAction
