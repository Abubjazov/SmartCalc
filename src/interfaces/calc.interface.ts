export enum CalcActionTypes {
	FETCH_CURRENT_STATE = 'FETCH_CURRENT_STATE',
	FETCH_CURRENT_STATE_SUCCESS = 'FETCH_CURRENT_STATE_SUCCESS',
	FETCH_CURRENT_STATE_ERROR = 'FETCH_CURRENT_STATE_ERROR',
}

export interface InputItems {
	key: string
	value: string
}

export interface UserState {
	items: number[]
	step: number
}

export interface CalcState {
	items: number[]
	inputItems: InputItems[]
	sortedItems: number[]
	sortDirection: boolean
	searchString: string
	status: 'waiting' | 'error' | 'loading'
	step: number
	error: null | string
}

interface FetchCurrentStateAction {
	type: CalcActionTypes.FETCH_CURRENT_STATE
}

interface FetchCurrentStateSuccessAction {
	type: CalcActionTypes.FETCH_CURRENT_STATE_SUCCESS
	payload: UserState
}

interface FetchCurrentStateErrorAction {
	type: CalcActionTypes.FETCH_CURRENT_STATE_ERROR
	payload: string
}

export type CalcAction =
	| FetchCurrentStateAction
	| FetchCurrentStateSuccessAction
	| FetchCurrentStateErrorAction
