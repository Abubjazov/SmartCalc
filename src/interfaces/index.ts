export * from './auth.interface'
export * from './calc.interface'
