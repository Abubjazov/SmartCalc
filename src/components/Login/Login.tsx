import { useEffect, useState } from 'react'
import { useActions } from '../../hooks/useActions'
import bcrypt from 'bcryptjs'

import { useTypedSelector } from '../../hooks/useTypedSelector'

import './Login.scss'

export const Login = (): JSX.Element => {
	const [submitStatus, setSubmitStatus] = useState<boolean>(true)

	const [email, setEmail] = useState<{
		value: string
		isValid: boolean | null
	}>({
		value: '',
		isValid: null,
	})
	const [password, setPassword] = useState<{
		value: string
		isValid: boolean | null
	}>({
		value: '',
		isValid: null,
	})

	const { status, error } = useTypedSelector(state => state.auth)
	const { fetchToken, clearError } = useActions()

	const setInputState = (value: string, field: string) => {
		let isValid = false

		if (field === 'email') {
			isValid =
				value.match(/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i) !== null
					? true
					: false
			setEmail({ value, isValid })
		}

		if (field === 'password') {
			isValid = value.length > 7
			setPassword({ value, isValid })
		}
	}

	useEffect(() => {
		setSubmitStatus(!(email.isValid === true && password.isValid === true))
	}, [email.isValid, password.isValid])

	useEffect(() => {
		let resetError = setTimeout(clearError, 3000)

		return () => {
			clearTimeout(resetError)
		}
	}, [clearError, error])

	return (
		<div className='login-widget'>
			<input
				className={
					email.isValid !== null ? (email.isValid ? 'null' : 'warning') : 'null'
				}
				type='email'
				placeholder='your@email.com'
				onChange={e => setInputState(e.target.value, 'email')}
				value={email.value}
			/>
			{email.isValid !== null ? (
				email.isValid ? null : (
					<p>Неверный формат email</p>
				)
			) : null}
			<input
				className={
					password.isValid !== null
						? email.isValid
							? 'null'
							: 'warning'
						: 'null'
				}
				type='password'
				placeholder='password'
				onChange={e => setInputState(e.target.value, 'password')}
				value={password.value}
			/>
			{password.isValid !== null ? (
				password.isValid ? null : (
					<p>Минимум 8 символов</p>
				)
			) : null}
			{status === 'error' && error?.includes('Authentication error!') ? (
				<p>Неверный логин или пароль</p>
			) : null}
			<button
				disabled={submitStatus}
				onClick={() =>
					fetchToken({
						email: email.value,
						password: bcrypt.hashSync(
							password.value,
							'$2a$10$9k5n.PAjQUFjv08.OlWN0e'
						),
					})
				}
			>
				Войти
			</button>
		</div>
	)
}
