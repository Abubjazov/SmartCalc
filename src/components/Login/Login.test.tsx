import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { AuthActionTypes } from '../../interfaces'

import { Login } from './Login'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

let store = mockStore({
	auth: {
		status: 'waiting',
		error: null,
	},
})

describe('Component: Login', () => {
	test('should render Login default', () => {
		const { asFragment } = render(
			<Provider store={store}>
				<Login />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should render Login with "valid login & password"', () => {
		const { asFragment } = render(
			<Provider store={store}>
				<Login />
			</Provider>
		)

		let inputItem = screen.getByPlaceholderText('your@email.com')
		userEvent.type(inputItem, 'test@test.com')
		expect(asFragment()).toMatchSnapshot()

		inputItem = screen.getByPlaceholderText('password')
		userEvent.type(inputItem, 'password')
		expect(asFragment()).toMatchSnapshot()

		inputItem = screen.getByPlaceholderText('your@email.com')
		userEvent.type(inputItem, '@.')
		expect(asFragment()).toMatchSnapshot()
	})

	test('should render Login status is "Error"', () => {
		store = mockStore({
			auth: {
				status: 'error',
				error: 'Пользователь не найден',
			},
		})
		const { asFragment } = render(
			<Provider store={store}>
				<Login />
			</Provider>
		)

		expect(asFragment()).toMatchSnapshot()
	})

	test('should dispatch an action on "Login" Button Click', () => {
		store = mockStore({
			auth: {
				status: 'waiting',
				error: null,
			},
		})

		const expectedActions = [
			{
				type: AuthActionTypes.FETCH_TOKEN,
				payload: {
					email: 'test@test.com',
					password:
						'$2a$10$9k5n.PAjQUFjv08.OlWN0eTRBKToGb2cen4PSPcja3PW8LsRYtVya',
				},
			},
		]

		render(
			<Provider store={store}>
				<Login />
			</Provider>
		)

		let inputItem = screen.getByPlaceholderText('your@email.com')
		userEvent.type(inputItem, 'test@test.com')

		inputItem = screen.getByPlaceholderText('password')
		userEvent.type(inputItem, 'password')

		userEvent.click(screen.getByText('Войти'))

		expect(store.getActions()).toEqual(expectedActions)
	})
})
