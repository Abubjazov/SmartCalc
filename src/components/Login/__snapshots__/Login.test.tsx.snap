// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Login should render Login default 1`] = `
<DocumentFragment>
  <div
    class="login-widget"
  >
    <input
      class="null"
      placeholder="your@email.com"
      type="email"
      value=""
    />
    <input
      class="null"
      placeholder="password"
      type="password"
      value=""
    />
    <button
      disabled=""
    >
      Войти
    </button>
  </div>
</DocumentFragment>
`;

exports[`Component: Login should render Login status is "Error" 1`] = `
<DocumentFragment>
  <div
    class="login-widget"
  >
    <input
      class="null"
      placeholder="your@email.com"
      type="email"
      value=""
    />
    <input
      class="null"
      placeholder="password"
      type="password"
      value=""
    />
    <p>
      Неверный логин или пароль
    </p>
    <button
      disabled=""
    >
      Войти
    </button>
  </div>
</DocumentFragment>
`;

exports[`Component: Login should render Login with "valid login & password" 1`] = `
<DocumentFragment>
  <div
    class="login-widget"
  >
    <input
      class="null"
      placeholder="your@email.com"
      type="email"
      value="test@test.com"
    />
    <input
      class="null"
      placeholder="password"
      type="password"
      value=""
    />
    <button
      disabled=""
    >
      Войти
    </button>
  </div>
</DocumentFragment>
`;

exports[`Component: Login should render Login with "valid login & password" 2`] = `
<DocumentFragment>
  <div
    class="login-widget"
  >
    <input
      class="null"
      placeholder="your@email.com"
      type="email"
      value="test@test.com"
    />
    <input
      class="null"
      placeholder="password"
      type="password"
      value="password"
    />
    <button>
      Войти
    </button>
  </div>
</DocumentFragment>
`;

exports[`Component: Login should render Login with "valid login & password" 3`] = `
<DocumentFragment>
  <div
    class="login-widget"
  >
    <input
      class="warning"
      placeholder="your@email.com"
      type="email"
      value="test@test.com@."
    />
    <p>
      Неверный формат email
    </p>
    <input
      class="warning"
      placeholder="password"
      type="password"
      value="password"
    />
    <button
      disabled=""
    >
      Войти
    </button>
  </div>
</DocumentFragment>
`;
